ocpi-endpoints-cpo-locations/src/main/scala/com/thenewmotion/ocpi/locations/CpoLocationsRoute.scala

```diff
@@ -31,50 +33,74 @@ class CpoLocationsRoute(
+
+  import msgs.v2_1.OcpiJsonProtocol._
+
+  implicit def locationsErrorResp(implicit errorMarshaller: ToResponseMarshaller[(StatusCode, ErrorResp)]): ToResponseMarshaller[LocationsError] = {
+      errorMarshaller.compose[LocationsError] { locationsError =>
+      val statusCode = locationsError match {
+        case (_: LocationNotFound | _: EvseNotFound | _: ConnectorNotFound) =>
+          NotFound // TODO: was OK?
+        case _ => InternalServerError
+      }
+      statusCode -> ErrorResp(GenericClientFailure, locationsError.reason.orElse(DefaultErrorMsg))
+    }
+  }
+
+  def route(apiUser: ApiUser)(implicit executionContext: ExecutionContext) =
+    handleRejections(AuthorizationRejectionHandler.Default) (routeWithoutRh(apiUser))
 
   private def limitToUse(clientLimit: Int) = Math.min(DefaultLimit, clientLimit)
 
-  private [locations] def routeWithoutRh(apiUser: ApiUser) = {
+  private[locations] def routeWithoutRh(apiUser: ApiUser)(implicit executionContext: ExecutionContext) = {
     get {
       pathEndOrSingleSlash {
         dateLimiters { (dateFrom: Option[DateTime], dateTo: Option[DateTime]) =>
           paged { (offset: Int, clientLimit: Int) =>
-            leftToRejection(service.locations(Pager(offset, limitToUse(clientLimit)),
-              dateFrom, dateTo)) { pagLocations =>
-
+            onSuccess(service.locations(
+              Pager(offset, limitToUse(clientLimit)), dateFrom, dateTo
+            )) { _.fold(complete(_), pagLocations => {
               val params = Map.empty ++
                 dateFrom.map(x => dateFromParam -> formatter.print(x)) ++
                 dateTo.map(x => dateToParam -> formatter.print(x))
 
-              respondWithPaginationHeaders( offset, limitToUse(clientLimit), params, pagLocations ) {
-                complete(SuccessWithDataResp(GenericSuccess, None, data = pagLocations.result))
+              respondWithPaginationHeaders(
+                offset, limitToUse(clientLimit), params, pagLocations
+              ) {
+                complete {
+                  SuccessWithDataResp(GenericSuccess, None, data = pagLocations.result)
+                }
               }
-            }
+            })}
           }
         }
       } ~
       pathPrefix(Segment) { locId =>
         pathEndOrSingleSlash {
-          leftToRejection(service.location(locId)) { location =>
-            complete(SuccessWithDataResp(GenericSuccess, None, data = location))
+          complete {
+            service.location(locId).map(_.map { location =>
+              SuccessWithDataResp(GenericSuccess, None, data = location)
+            })
           }
         } ~
         pathPrefix(Segment) { evseId =>
           pathEndOrSingleSlash {
-            leftToRejection(service.evse(locId, evseId)) { evse =>
-              complete(SuccessWithDataResp(GenericSuccess, None, data = evse))
+            complete {
+              service.evse(locId, evseId).map(_.map { evse =>
```

Comment:
Might be nice to reduce some boilerplate even more here with some pimping...

In `com.thenewmotion.ocpi.package`
```scala
  implicit class PimpedFutureDisj[L, R](value: Future[L \/ R]) {
    def mapRight[T](f: R => T) = value.map(_.map(f))
  }
```
then you can do 
```scala
service.location(locId).mapRight { location =>
  SuccessWithDataResp(GenericSuccess, None, data = location)
}
```